## Classification using Deep Learning
### Amir, 2025, NUS SoC ACE
#### .
#### We use the heart disease dataset with 2 classes.


In [82]:
# first neural network with keras tutorial
from numpy import loadtxt
from sklearn.preprocessing import MinMaxScaler
from keras.models import Sequential
from keras.layers import Dense , Dropout
import numpy as np

In [2]:
# mounting the Google drive
#from google.colab import drive

#drive.mount('/content/drive')

# Load the dataset
fp = "sample_data/datasets_heart_noheader.csv"

In [92]:
# load the dataset
dataset = loadtxt(fp, delimiter=',')
# split into input (X) and output (y) variables
print(dataset.shape)
scaler_x = MinMaxScaler()
scaler_y = MinMaxScaler()
X = dataset[0:270,0:13]
print(X.shape)
y = dataset[0:270,13]
X2 = dataset[270:302,0:13]
y2 = dataset[270:302,13]
X_scaled = scaler_x.fit_transform(X)
X2_scaled = scaler_x.fit_transform(X2)
y_scaled = scaler_y.fit_transform(y.reshape(-1,1))
y2_scaled = scaler_y.fit_transform(y2.reshape(-1,1))
print(X2_scaled)
print(y2_scaled)

(303, 14)
(270, 13)
[[0.63888889 1.         0.         0.06666667 0.88625592 0.
  1.         0.54347826 0.         1.         0.         0.75
  0.        ]
 [0.63888889 0.         0.         1.         0.44549763 1.
  0.         0.60869565 1.         0.63636364 0.5        0.5
  0.        ]
 [0.88888889 1.         0.66666667 0.7        0.38388626 0.
  0.         0.65217391 0.         0.18181818 0.5        0.
  1.        ]
 [0.25       1.         0.         0.16666667 0.18009479 0.
  0.5        0.58695652 1.         0.63636364 0.         0.
  0.        ]
 [0.77777778 1.         0.         0.5        0.26540284 0.
  0.         0.58695652 1.         0.90909091 1.         0.5
  1.        ]
 [0.77777778 0.         0.         0.23333333 0.31279621 0.
  0.5        0.5        1.         0.         0.5        0.
  0.5       ]
 [0.66666667 1.         0.         0.9        0.21327014 1.
  0.         0.         0.         0.22727273 0.5        0.5
  0.        ]
 [0.61111111 0.         0.         0.

In [93]:
# define the keras model
from tensorflow.keras import regularizers
model = Sequential()
model.add(Dense(12, input_dim=13, activation='relu'))
#model.add(Dense(4, activation='relu'))
#model.add(Dropout(0.5))
#model.add(Dense(4, activation='relu', kernel_regularizer=regularizers.l2(0.01)))
#model.add(Dropout(0.5))
model.add(Dense(4, activation='relu'))
model.add(Dense(1, activation='sigmoid'))

/usr/local/lib/python3.12/dist-packages/keras/src/layers/core/dense.py:93: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


In [94]:
# compile the keras model
model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])

In [95]:
# fit the keras model on the dataset
model.fit(X_scaled, y_scaled, epochs=600, batch_size=10)

Epoch 1/600
27/27 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - accuracy: 0.3545 - loss: 0.7201
Epoch 2/600
27/27 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.5484 - loss: 0.6858
Epoch 3/600
27/27 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.5440 - loss: 0.6722
Epoch 4/600
27/27 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.5557 - loss: 0.6538
Epoch 5/600
27/27 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.5279 - loss: 0.6469
Epoch 6/600
27/27 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.5408 - loss: 0.6392
Epoch 7/600
27/27 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.5578 - loss: 0.6239
Epoch 8/600
27/27 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.6496 - loss: 0.5976
Epoch 9/600
27/27 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.6369 - loss: 0.6007
Epoch 10/600
27/27 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.6673 - loss: 0.5910
Epoch 11/600
27/27 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.6962 - loss: 0.5663
Epoch 12/600
27/27 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy:

In [96]:
# evaluate the keras model
results = model.evaluate(X_scaled, y_scaled, verbose=2)
print('Accuracy: %.2f' % (results[1]*100))

9/9 - 1s - 57ms/step - accuracy: 0.9370 - loss: 0.1635
Accuracy: 93.70


In [101]:
# make probability predictions with the model
predictions = model.predict(X_scaled)
# round predictions
rounded = [round(x[0]) for x in predictions]

9/9 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


In [102]:
# make class predictions with the model
for i in range(15):
	print('%s => %d (expected %d)' % (X[i].tolist(), rounded[i], y[i]))
  #print('predicted %d (expected %d)' % (rounded[i], y[i]))

[63.0, 1.0, 3.0, 145.0, 233.0, 1.0, 0.0, 150.0, 0.0, 2.3, 0.0, 0.0, 1.0] => 1 (expected 1)
[37.0, 1.0, 2.0, 130.0, 250.0, 0.0, 1.0, 187.0, 0.0, 3.5, 0.0, 0.0, 2.0] => 1 (expected 1)
[41.0, 0.0, 1.0, 130.0, 204.0, 0.0, 0.0, 172.0, 0.0, 1.4, 2.0, 0.0, 2.0] => 1 (expected 1)
[56.0, 1.0, 1.0, 120.0, 236.0, 0.0, 1.0, 178.0, 0.0, 0.8, 2.0, 0.0, 2.0] => 1 (expected 1)
[57.0, 0.0, 0.0, 120.0, 354.0, 0.0, 1.0, 163.0, 1.0, 0.6, 2.0, 0.0, 2.0] => 1 (expected 1)
[57.0, 1.0, 0.0, 140.0, 192.0, 0.0, 1.0, 148.0, 0.0, 0.4, 1.0, 0.0, 1.0] => 1 (expected 1)
[56.0, 0.0, 1.0, 140.0, 294.0, 0.0, 0.0, 153.0, 0.0, 1.3, 1.0, 0.0, 2.0] => 1 (expected 1)
[44.0, 1.0, 1.0, 120.0, 263.0, 0.0, 1.0, 173.0, 0.0, 0.0, 2.0, 0.0, 3.0] => 1 (expected 1)
[52.0, 1.0, 2.0, 172.0, 199.0, 1.0, 1.0, 162.0, 0.0, 0.5, 2.0, 0.0, 3.0] => 1 (expected 1)
[57.0, 1.0, 2.0, 150.0, 168.0, 0.0, 1.0, 174.0, 0.0, 1.6, 2.0, 0.0, 2.0] => 1 (expected 1)
[54.0, 1.0, 0.0, 140.0, 239.0, 0.0, 1.0, 160.0, 0.0, 1.2, 2.0, 0.0, 2.0] => 1 (expected 1)

In [103]:
M= len(rounded)
print( 'Accuracy= ' , 1- np.sum(np.abs(rounded - y )) / M)

Accuracy=  0.937037037037037


In [105]:
# evaluate the keras model, testing
results = model.evaluate(X2_scaled, y2_scaled)
print('Accuracy: %.2f' % (results[1]*100))

1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 32ms/step - accuracy: 0.8125 - loss: 0.8035
Accuracy: 81.25


In [106]:
# make probability predictions with the model, testing
predictions = model.predict(X2_scaled)
# round predictions
rounded = [round(x[0]) for x in predictions]

1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 26ms/step


In [108]:
# make class predictions with the model, testing
for i in range(25):
 print('%s => %d (expected %d)' % (X2[i].tolist(), rounded[i], y2[i]))

[58.0, 1.0, 0.0, 114.0, 318.0, 0.0, 2.0, 140.0, 0.0, 4.4, 0.0, 3.0, 1.0] => 0 (expected 0)
[58.0, 0.0, 0.0, 170.0, 225.0, 1.0, 0.0, 146.0, 1.0, 2.8, 1.0, 2.0, 1.0] => 0 (expected 0)
[67.0, 1.0, 2.0, 152.0, 212.0, 0.0, 0.0, 150.0, 0.0, 0.8, 1.0, 0.0, 3.0] => 0 (expected 0)
[44.0, 1.0, 0.0, 120.0, 169.0, 0.0, 1.0, 144.0, 1.0, 2.8, 0.0, 0.0, 1.0] => 0 (expected 0)
[63.0, 1.0, 0.0, 140.0, 187.0, 0.0, 0.0, 144.0, 1.0, 4.0, 2.0, 2.0, 3.0] => 0 (expected 0)
[63.0, 0.0, 0.0, 124.0, 197.0, 0.0, 1.0, 136.0, 1.0, 0.0, 1.0, 0.0, 2.0] => 1 (expected 0)
[59.0, 1.0, 0.0, 164.0, 176.0, 1.0, 0.0, 90.0, 0.0, 1.0, 1.0, 2.0, 1.0] => 0 (expected 0)
[57.0, 0.0, 0.0, 140.0, 241.0, 0.0, 1.0, 123.0, 1.0, 0.2, 1.0, 0.0, 3.0] => 0 (expected 0)
[45.0, 1.0, 3.0, 110.0, 264.0, 0.0, 1.0, 132.0, 0.0, 1.2, 1.0, 0.0, 3.0] => 0 (expected 0)
[68.0, 1.0, 0.0, 144.0, 193.0, 1.0, 1.0, 141.0, 0.0, 3.4, 1.0, 2.0, 3.0] => 0 (expected 0)
[57.0, 1.0, 0.0, 130.0, 131.0, 0.0, 1.0, 115.0, 1.0, 1.2, 1.0, 1.0, 3.0] => 0 (expected 0)


In [109]:
M= len(rounded)
print( 'Accuracy= ' , 1- np.sum(np.abs(rounded - y2)) / M)

Accuracy=  0.8125


## Practices for You
1. Run the program a few times. Is there any meaningful difference between the results? Why?
  - Yes, even if the neural network (NN) architecture is not changed and the dataset is exactly the same, running the training process again can and often will change the model performance.
While the evaluation of a pre-trained, frozen model on the same data should produce identical results, the training process itself is stochastic (non-deterministic).
Here is a breakdown of why this happens and how to manage it:
Why Results Change (Sources of Randomness)
Random Weight Initialization: Neural networks start by initializing their internal weights with random numbers. Different initializations lead to different paths during optimization, resulting in different final models.
Data Shuffling: Data is usually shuffled at the start of each epoch to improve learning. If not strictly controlled, this means the model sees examples in a different order each time, affecting weight updates.
GPU Non-determinism: Many GPU operations (e.g., in CUDA/PyTorch/TensorFlow) are parallelized and not strictly deterministic. The order of operations may vary slightly, leading to different floating-point results that accumulate over training.
Hardware/Numerical Differences: Small differences in how CPUs or GPUs handle floating-point math can result in different, yet equally valid, performance metrics.
2. Train the model for 1100 epochs. what are the training and testing accuracy?
  - Training - 86.3, Testing - 75
3. repeat that with 6000 training epochs. what are the training and testing accuracy?
- Training - 90, Testing - 75
4. move back to 600 training epochs. add two dense layers with optional features. what are the training and testing accuracy?
  - adding more layers and optional features such as dropout amd regularizer in this case did not improve accuracy. Training accuracy is 82%
5. go for 6000 training epochs. Do you feel any overfitting?
6. set back everything to the original setting, then try to normalize the training and testing features and outputs using standard_scaler. Then see if anything gets better or not.
 - By normalizing the data using MinMaxScaler, the performance improved. Trainign accuracy is 93.7 and testing accuracy is 81.25

#### points:
- Use GPU as your runtime type